File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Effect/Combat/WebLineEffect.ts
import { FairyGUI } from "csharp";
import { Mathf } from "../../../../../Core/Module/Math/Mathf";
import Color from "../../DataDefine/Color";
import FileHelper from "../../FileHelper";
import UIHelper from "../../UI/UIHelper";
import { AbstractEffect } from "../AbstractEffect";


export class WebLineEffect extends AbstractEffect{
    public Loader:FairyGUI.GLoader;
    private _Scale:number;
    public constructor(private _X:number,private _Y:number,facingLeft:boolean) {
        super();
        this.Loader = UIHelper.CreateGLoader();
        this.Loader.url = FileHelper.FormatPath("Vfx/horizontal_line.png");
        this.Loader.SetPosition(_X + Mathf.RandomRange(-20,20),_Y * 128 + Mathf.RandomRange(-10,10), 0);
        this.Loader.rotation = Mathf.RandomRange(170,185);
        if(!facingLeft)
            this.Loader.rotation += 180;
        this._Scale = Mathf.RandomRange(0.8,1);
        this.Loader.scaleY = this._Scale;
        let g = Mathf.RandomRange(153, 230);
        let color = new Color(g,g,g + 26);
        this.Loader.alpha = 0;
        this.Loader.color = color.UnityColor();
        this.Loader.TweenFade(0.8,0.5).SetEase(FairyGUI.EaseType.SineInOut);
        this.Loader.TweenFade(0.01,0.5).SetEase(FairyGUI.EaseType.QuartOut).SetDelay(0.5);
        this.Loader.blendMode = FairyGUI.BlendMode.Add;
    }
    
    public Update() {
        if(this.Duration >= 1){
            this.Loader.Dispose();
            this.IsDone = true;
            return;
        }
        this.Loader.scaleX = this._Scale * 2 * Mathf.Cos(1 - this.Duration * 16) / 4 + 1.5;
        super.Update();
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Effect/Combat/WebLineEffect.cs
using FairyGUI;
using UnityEngine;

namespace ModTheSpire.Effect.Combat
{
    public class WebLineEffect : AbstractEffect
    {
        private float mX;
        private float mY;
        private GLoader mLoader;
        private float mScale;
        public WebLineEffect(float x, float y,bool facingLeft)
        {
            mX = x;
            mY = y;
            mLoader = (GLoader)UIObjectFactory.NewObject(ObjectType.Loader);
            mLoader.url = "Vfx/horizontal_line.png";
            mLoader.SetPosition(x + Random.Range(-20, 20), y * 128 + Random.Range(-10, 10), 0);
            mLoader.rotation = Random.Range(170,185);
            if(!facingLeft)
                mLoader.rotation += 180;
            mScale = Random.Range(0.8f,1f);
            mLoader.scaleY = mScale;
            var g = Random.Range(153, 230);
            var color = new Color(g,g,g + 26);
            mLoader.alpha = 0;
            mLoader.color = color;
            mLoader.TweenFade(0.8f,0.5f).SetEase(EaseType.SineInOut);
            mLoader.TweenFade(0.01f,0.5f).SetEase(EaseType.QuartOut).SetDelay(0.5f);
            mLoader.blendMode = BlendMode.Add;
        }

        public override void Update()
        {
            base.Update();
            mLoader.scaleX = this.mScale * 2 * Mathf.Cos(1 - this.Duration * 16) / 4 + 1.5f;
            if(this.Duration >= 1)
                Finish();
        }

        public override void Dispose()
        {
            base.Dispose();
            mLoader.Dispose();
        }
    }
}